Rover: Add a line-following mode [WIP]

rishabsingh3003 opened the pull request:
This PR is in response to #7590.

- I have attempted to simulate Analog IR Sensor's [https://www.sparkfun.com/products/9453]. These sensors are widely used in line following robots - white surfaces reflect much more light than black, so, when directed towards a white surface, the voltage output will be lower than that on a black surface. A minimum of two of these sensors are required on the far edges of the robot to determine where the vehicle is with respect to the black line.  
![image](https://user-images.githubusercontent.com/36970042/71436639-58bf9080-2714-11ea-956f-aaa6cdcfb37e.png)


- A new IR Sensor library is created which is divided into front-end, back-end, and an analog driver similar to other standard Ardupilot sensor libraries. This will make it easier to add other line following sensors to the library. Certain parameters are also defined which need to be specified by the user. Using the values that are picked up from the sensors, an error is generated to approximate the position of the vehicle with respect to the line. **The first 8 commits include these additions**

- A new "Line Follower" mode is created in the **last 10 commits**. The IR sensors are read periodically and the generated error is directly fed to ` AR_AttitudeControl::get_steering_out_rate()`.
The following image explains the theory behind how this mode works. 

![image](https://user-images.githubusercontent.com/36970042/71437234-afc66500-2716-11ea-9a4d-3cd783828564.png)

- **TESTING:**

1. The simulation in SITL works well. To simulate, first change the param's :
 `param set IR_PIN_1 16`
`param set IR_PIN_2 17`
`param set IR_ADC_VOLTAGE 5`

Then write the commands: 
`mode 17`
`arm throttle`
`rc 3 1700`

The GIF of a simulation is posted below:
![Line_Follower_SITL](https://user-images.githubusercontent.com/36970042/71438414-b4414c80-271b-11ea-9fa3-d29cb8ac206d.gif)

2. `Tools/scripts/build_all.sh` script was run. All systems were built successfully. 

3. I had attached the analog sensors to the 3.3V ADC of my Pixhawk, The values were read correctly. 

4. Unfortunately, I do not have a compatible motor driver (ESC) to build the rover, therefore I **can not test the mode on an actual vehicle** but this PR should at least give a good start to this age-long issue. I had tested a similar algorithm on an Arduino Uno based vehicle with a cheap motor driver and the mode worked well.

- @peterbarker please tell me the modifications required in this PR, if they are needed. I realize there are a lot of commits! Hopefully, this will be a start to Ardupilot's Line Follower vehicles and other developers can contribute to this.

khancyr commented on libraries/AP_HAL_SITL/sitl_irsensor.cpp:
```diff
@@ -0,0 +1,72 @@
+
+    if (crosstrack_error_1>= max_sensor_range){ //force sensor to stay in range limit
+        crosstrack_error_1 = max_sensor_range;
+    }if(crosstrack_error_2>= max_sensor_range){
```
Syntax issue here

khancyr commented on libraries/AP_HAL_SITL/sitl_irsensor.cpp:
```diff
@@ -0,0 +1,72 @@
+  This simulates a Analog IR Sensor
+
+ */
+
```
This file need a reformat asp plenties space and float qualifier are missing

khancyr commented on libraries/AR_IR/AR_IR_Sensor.cpp:
```diff
@@ -0,0 +1,47 @@
+	drivers->read();
+}
+
+void AR_IR:: calc_line_error(float &line_error){
```
The file also need a reformat

khancyr commented on libraries/AR_IR/AR_IR_Sensor.h:
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#define IR_1_PIN_NUM  16
```
Those don't sems to be used

khancyr replied:
Ha I saw the usage...

rishabsingh3003 commented:
@khancyr Thanks for the quick review!
As I understand your suggestions, you want me to make a backend for SITL instead of changing HAL files itself. That makes a lot of sense. I'll fix the formatting issues too. Will take some time, though.

khancyr commented:
Yes, I will confirm with @peterbarker as I think it will be simplier

rishabsingh3003 commented:
@khancyr I dropped all the HAL_SITL commits and moved them to IR backend instead as you asked (Therefore HAL is untouched now). I have formatted the new files as well. Hope they are okay this time. Do let me know if any changes are required.

rishabsingh3003 commented:
This probably belongs as a Lua script now! The implemented algorithm seemed okay, so if anyone wants to pick on this in the future, please go ahead! I am closing this PR.
